base/applications/taskmgr/procpage.c

```diff
@@ -931,3 +936,300 @@ int CALLBACK ProcessPageCompareFunc(LPARAM lParam1, LPARAM lParam2, LPARAM lPara
     }
     return ret;
 }
+
+static BOOL DevicePathToDosPath(LPWSTR lpPath, DWORD dwLength)
+{
+    WCHAR *pszPath;
+    WCHAR cDrive;
+    BOOL bSuccess = FALSE;
+
+    /* Check if lpPath is a device path */
+    if (_wcsnicmp(lpPath, L"\\Device\\", 8) != 0)
+    {
+        return FALSE;
+    }
+
+    pszPath = HeapAlloc(GetProcessHeap(), 0, dwLength * sizeof(WCHAR));
+
+    if (!pszPath)
+    {
+        return FALSE;
+    }
+
+    for (cDrive = L'A'; cDrive <= L'Z'; cDrive++)
+    {
+        WCHAR szDrive[3];
+        WCHAR szDevPath[MAX_PATH];
+
+        szDrive[0] = cDrive;
+        szDrive[1] = L':';
+        szDrive[2] = L'\0';
+
+        if (QueryDosDeviceW(szDrive, szDevPath, _countof(szDevPath)) != 0)
+        {
+            size_t len = wcslen(szDevPath);
+            bSuccess = (_wcsnicmp(lpPath, szDevPath, len) == 0);
+
+            if (bSuccess)
+            {
+                StringCchPrintfW(pszPath, dwLength, L"%s%s", szDrive, lpPath + len);
+                break;
+            }
+        }
+    }
+
+    if (bSuccess)
+    {
+        StringCchCopyW(lpPath, dwLength, pszPath);
+    }
+
+    HeapFree(GetProcessHeap(), 0, pszPath);
+
+    return bSuccess;
+}
+
+static DWORD GetProcessExecutablePath(HANDLE hProcess, LPWSTR lpExePath, DWORD dwLength)
+{
+    BYTE StaticBuffer[sizeof(UNICODE_STRING) + (MAX_PATH * sizeof(WCHAR))];
+    PVOID DynamicBuffer = NULL;
+    PUNICODE_STRING ImagePath = NULL;
+    WCHAR *pszExePath = NULL;
+    ULONG SizeNeeded;
+    NTSTATUS Status;
+    DWORD dwRet = 0;
+
+    Status = NtQueryInformationProcess(hProcess,
+                                       ProcessImageFileName,
+                                       StaticBuffer,
+                                       sizeof(StaticBuffer) - sizeof(WCHAR),
+                                       &SizeNeeded);
+
+    if (Status == STATUS_INFO_LENGTH_MISMATCH)
+    {
+        DynamicBuffer = HeapAlloc(GetProcessHeap(), 0, SizeNeeded + sizeof(WCHAR));
+
+        if (!DynamicBuffer)
+        {
+            return 0;
+        }
+
+        Status = NtQueryInformationProcess(hProcess,
+                                           ProcessImageFileName,
+                                           DynamicBuffer,
+                                           SizeNeeded,
+                                           &SizeNeeded);
+
+        ImagePath = (PUNICODE_STRING)DynamicBuffer;
+    }
+    else
+    {
+        ImagePath = (PUNICODE_STRING)StaticBuffer;
+    }
+
+    if (!NT_SUCCESS(Status))
+    {
+        goto Cleanup;
+    }
+
+    pszExePath = HeapAlloc(GetProcessHeap(), 0, ImagePath->Length + sizeof(WCHAR));
+
+    if (!pszExePath)
+    {
+        goto Cleanup;
+    }
+
+    memcpy(pszExePath, ImagePath->Buffer, ImagePath->Length);
+    pszExePath[ImagePath->Length / sizeof(WCHAR)] = UNICODE_NULL;
```

Comment:
```suggestion
    StringCbCopyNW(pszExePath, ImagePath->Length + sizeof(WCHAR), ImagePath->Buffer, ImagePath->Length);
```

Reply:
Is it safe to use `StringCbCopyNW` with `UNICODE_STRING` ? The documentation says that for `StringCbCopyNW` the source string must be null-terminated.

Reply:
yes but actually you specify what the actual number of characters is in the source string, so it doesn't matter.